File: apps/api/app/archive_knowledge/contracts/policy.py
from __future__ import annotations

from pydantic import BaseModel, Field

from app.archive_knowledge.contracts.common import ArtifactRef, P1LifecycleStatus
from app.archive_knowledge.contracts.rule import RuleContract


class StageExecutionContract(BaseModel):
    stage_id: str
    stage_name: str
    enabled: bool = True
    order_hint: int
    input_artifacts: list[ArtifactRef] = Field(default_factory=list)
    output_artifacts: list[ArtifactRef] = Field(default_factory=list)
    rule_ids: list[str] = Field(default_factory=list)
    entry_contract_id: str | None = None
    exit_contract_id: str | None = None
    can_run_independently: bool = False
    downstream_stage_ids: list[str] = Field(default_factory=list)


class PolicyPackageVersion(BaseModel):
    policy_package_version_id: str
    version_label: str
    status: P1LifecycleStatus
    hash: str
    created_at: str
    previous_policy_package_version_id: str | None = None
    stage_contracts: list[StageExecutionContract] = Field(default_factory=list)
    rule_contracts: list[RuleContract] = Field(default_factory=list)
    compatible_output_contracts: list[str] = Field(default_factory=list)


class PolicyPackage(BaseModel):
    policy_package_id: str
    policy_package_name: str
    business_domain: str
    knowledge_types: list[str] = Field(default_factory=list)
    owner: str
    lifecycle_status: P1LifecycleStatus
    current_version_id: str
    versions: list[PolicyPackageVersion] = Field(default_factory=list)


class PolicyRuntimeSnapshot(BaseModel):
    snapshot_id: str
    archive_id: str
    run_id: str
    policy_package_id: str
    policy_package_version_id: str
    policy_package_version_hash: str
    frozen_at: str
    stage_contract_refs: list[str] = Field(default_factory=list)
    rule_contract_refs: list[str] = Field(default_factory=list)

File: apps/web/src/features/p1/contracts/policy.ts
import type { P1ArtifactRef, P1LifecycleStatus } from "./common";
import type { RuleContract } from "./rule";

export interface StageExecutionContract {
  stage_id: string;
  stage_name: string;
  enabled: boolean;
  order_hint: number;
  input_artifacts: P1ArtifactRef[];
  output_artifacts: P1ArtifactRef[];
  rule_ids: string[];
  entry_contract_id?: string;
  exit_contract_id?: string;
  can_run_independently: boolean;
  downstream_stage_ids: string[];
}

export interface PolicyPackageVersion {
  policy_package_version_id: string;
  version_label: string;
  status: P1LifecycleStatus;
  hash: string;
  created_at: string;
  previous_policy_package_version_id?: string;
  stage_contracts: StageExecutionContract[];
  rule_contracts: RuleContract[];
  compatible_output_contracts: string[];
}

export interface PolicyPackage {
  policy_package_id: string;
  policy_package_name: string;
  business_domain: string;
  knowledge_types: string[];
  owner: string;
  lifecycle_status: P1LifecycleStatus;
  current_version_id: string;
  versions: PolicyPackageVersion[];
}

export interface PolicyRuntimeSnapshot {
  snapshot_id: string;
  archive_id: string;
  run_id: string;
  policy_package_id: string;
  policy_package_version_id: string;
  policy_package_version_hash: string;
  frozen_at: string;
  stage_contract_refs: string[];
  rule_contract_refs: string[];
}
